### ISSUE TITLE
fp16 tensor.nnet.corr.CorrMM error on CPU

### ISSUE THREAD

**Issue by astooke** [Theano/Theano#5867]
Hi,

 I have a simple 2d convnet I'm trying to run in FP16 on GPU.  But as part of the program I instantiate the network on the CPU, basically just to do one forward pass to see what comes out.  Can definitely get around this but anyway it broke entirely:

```
Disabling C code for Elemwise{mul,no_inplace} due to unsupported float16
Disabling C code for CorrMM{valid, (2, 2), (1, 1)} due to unsupported float16
Disabling C code for Elemwise{Composite{(i0 * ((i1 + i2) + Abs((i1 + i2))))}}[(0, 1)] due to unsupported float16
Disabling C code for CorrMM{valid, (2, 2), (1, 1)} due to unsupported float16
Disabling C code for Elemwise{Composite{(i0 * ((i1 + i2) + Abs((i1 + i2))))}}[(0, 1)] due to unsupported float16
Disabling C code for Dot22 due to unsupported float16
Disabling C code for Elemwise{Composite{(i0 * ((i1 + i2) + Abs((i1 + i2))))}}[(0, 1)] due to unsupported float16
Disabling C code for Dot22 due to unsupported float16
Disabling C code for SoftmaxWithBias due to unsupported float16
Disabling C code for Elemwise{sub,no_inplace} due to unsupported float16
Disabling C code for Elemwise{Composite{((i0 * i1) + i2)}}[(0, 1)] due to unsupported float16
Disabling C code for CorrMM{valid, (2, 2), (1, 1)} due to unsupported float16
Disabling C code for Elemwise{Composite{(i0 * ((i1 + i2) + Abs((i1 + i2))))}}[(0, 1)] due to unsupported float16
Disabling C code for CorrMM{valid, (2, 2), (1, 1)} due to unsupported float16
Disabling C code for Elemwise{Composite{(i0 * ((i1 + i2) + Abs((i1 + i2))))}}[(0, 1)] due to unsupported float16
Disabling C code for Dot22 due to unsupported float16
Disabling C code for Elemwise{Composite{(i0 * ((i1 + i2) + Abs((i1 + i2))))}}[(0, 1)] due to unsupported float16
Traceback (most recent call last):
  File "/home/adam/GitRepos/Theano/theano/compile/function_module.py", line 884, in __call__
    self.fn() if output_subset is None else\
  File "/home/adam/GitRepos/Theano/theano/gof/op.py", line 888, in rval
    r = p(n, [x[0] for x in i], o)
  File "/home/adam/GitRepos/Theano/theano/gof/op.py", line 778, in perform
    "Did you used Theano flags mode=FAST_COMPILE?"
theano.gof.utils.MethodNotDefined: ('perform', <class 'theano.tensor.nnet.corr.CorrMM'>, 'CorrMM', 'Did you used Theano flags mode=FAST_COMPILE? You can use optimizer=fast_compile instead.')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "sandbox/adam/synkhronos/launch_scripts/atari_rec_m_gpu_samp.py", line 30, in <module>
    n_gpu=6, n_per_gpu=5, n_recording=20, buf_size=max_path_len, seed=1, reset_release=30)
  File "/home/adam/GitRepos/rllab-private/sandbox/adam/gpu_sampler/rec_m_gpu_parallel_sampler.py", line 118, in initialize
    par_objs = build_par_objs(env, policy, n_gpu, n_per_gpu, n_recording, buf_size)
  File "/home/adam/GitRepos/rllab-private/sandbox/adam/gpu_sampler/rec_m_gpu_parallel_sampler.py", line 437, in build_par_objs
    a, agent_info = policy.get_action(o)
  File "/home/adam/GitRepos/rllab-private/rllab/policies/categorical_conv_policy.py", line 125, in get_action
    prob = self._f_prob([flat_obs])[0]
  File "/home/adam/GitRepos/Theano/theano/compile/function_module.py", line 898, in __call__
    storage_map=getattr(self.fn, 'storage_map', None))
  File "/home/adam/GitRepos/Theano/theano/gof/link.py", line 325, in raise_with_op
    reraise(exc_type, exc_value, exc_trace)
  File "/home/adam/anaconda3/envs/rllab_dev/lib/python3.5/site-packages/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/home/adam/GitRepos/Theano/theano/compile/function_module.py", line 884, in __call__
    self.fn() if output_subset is None else\
  File "/home/adam/GitRepos/Theano/theano/gof/op.py", line 888, in rval
    r = p(n, [x[0] for x in i], o)
  File "/home/adam/GitRepos/Theano/theano/gof/op.py", line 778, in perform
    "Did you used Theano flags mode=FAST_COMPILE?"
theano.gof.utils.MethodNotDefined: ('perform', <class 'theano.tensor.nnet.corr.CorrMM'>, 'CorrMM', 'Did you used Theano flags mode=FAST_COMPILE? You can use optimizer=fast_compile instead.')
Apply node that caused the error: CorrMM{valid, (2, 2), (1, 1)}(InplaceDimShuffle{0,x,1,2}.0, Subtensor{::, ::, ::int64, ::int64}.0)
Toposort index: 13
Inputs types: [TensorType(float16, (False, True, False, False)), TensorType(float16, (False, True, False, False))]
Inputs shapes: [(1, 1, 105, 80), (16, 1, 4, 4)]
Inputs strides: [(16800, 16800, 160, 2), (32, 32, -8, -2)]
Inputs values: ['not shown', 'not shown']
Outputs clients: [[Elemwise{Composite{(i0 * ((i1 + i2) + Abs((i1 + i2))))}}[(0, 1)](TensorConstant{(1, 1, 1, 1) of 0.5}, CorrMM{valid, (2, 2), (1, 1)}.0, InplaceDimShuffle{x,0,x,x}.0)]]

Backtrace when the node is created(use Theano flag traceback.limit=N to make it longer):
  File "sandbox/adam/synkhronos/launch_scripts/atari_rec_m_gpu_samp.py", line 30, in <module>
    n_gpu=6, n_per_gpu=5, n_recording=20, buf_size=max_path_len, seed=1, reset_release=30)
  File "/home/adam/GitRepos/rllab-private/sandbox/adam/gpu_sampler/rec_m_gpu_parallel_sampler.py", line 117, in initialize
    policy = policy_cls(**policy_args)
  File "/home/adam/GitRepos/rllab-private/rllab/policies/categorical_conv_policy.py", line 73, in __init__
    nn_prob = L.get_output(prob_network.output_layer)
  File "/home/adam/GitRepos/Lasagne/lasagne/layers/helper.py", line 190, in get_output
    all_outputs[layer] = layer.get_output_for(layer_inputs, **kwargs)
  File "/home/adam/GitRepos/Lasagne/lasagne/layers/conv.py", line 331, in get_output_for
    conved = self.convolve(input, **kwargs)
  File "/home/adam/GitRepos/Lasagne/lasagne/layers/conv.py", line 609, in convolve
    filter_flip=self.flip_filters)
  File "/home/adam/GitRepos/rllab-private/rllab/core/network.py", line 24, in wrapped_conv
    return theano.tensor.nnet.conv2d(*args, **kwargs)

```

**Comment by nouiz** [comment]
We don't support float16 on the CPU, except with Python code. I added a
quick comment in

https://github.com/Theano/Theano/pull/5865

to don't introduce the nodes that don't support it in this case, but it
will happen elsewhere too.

If you or someone else want to make some implementation of it, it would be
possible to modify the CorMM* ops to call the float32 gemm(BLAS don't
support a float16 gemm). In the im2col op, it could upcast the float16 to
float32 for the computation. But this will upcast the output, so in the opt
that introduce them, we could add a cast in the graph that will convert it
back to float16. (and it would need an update of the make_node to tell the
output will be in float32).

float16 on CPU isn't our focus. It don't give computation speed up and the
memory usage isn't as limited as on the GPU.

**Comment by astooke** [comment]
Thanks for the response!

To be clear, personally, I'm fine with a Python-only implementation of this piece, I just want it to run once :)

**Comment by lamblin** [comment]
Fixed in #5865.
